python/hail/methods/misc.py

```diff
@@ -185,6 +185,61 @@ def rename_duplicates(dataset, name='unique_id'):
            intervals=oneof(Interval, listof(Interval)),
            keep=bool)
 def filter_intervals(ds, intervals, keep=True):
+    """Filter rows with an interval or list of intervals.
+
+    .. note::
+
+    Requires the dataset to have a partition key of type :class:`.TLocus`.
+
+    Examples
+    --------
+
+    Filter to loci falling within one interval:
+
+    >>> ds_result = hl.filter_intervals(dataset, Interval.parse('17:38449840-38530994'))
+
+    Remove all loci within list of intervals:
+
+    >>> intervals = [Interval.parse(x) for x in ['1:50M-75M', '2:START-400000', '3-22']]
+    >>> ds_result = vds.filter_intervals(intervals)
+
+    Notes
+    -----
+    This method takes an argument of :class:`.Interval` or list of
+    :class:`.Interval`.
+
+    Based on the ``keep`` argument, this method will either restrict to loci in
+    the supplied interval ranges, or remove all loci in those ranges.  Note that
+    intervals are left-inclusive, and right-exclusive; the below interval
+    includes the locus ``15:100000`` but not ``15:101000``.
+
+    >>> interval = Interval.parse('15:100000-101000')
+
+    When ``keep=True``, partitions that don't overlap any supplied interval
+    will not be loaded at all.  This enables ``filter_intervals`` to be used for
+    reasonably low-latency queries of small ranges of the genome, even on large
+    datasets.
+
+    Parameters
+    ----------
+    intervals : :class:`.Interval` or :obj:`list` of :class:`.Interval`
+        Interval(s) to filter on.
+    keep : :obj:`bool`
+        If ``True``, keep only loci fall within any interval in `intervals`. If
+        ``False``, keep only loci that fall outside all intervals in
+        `intervals`.
+
+    Returns
+    -------
+    :class:`.MatrixTable`
+    """
+
+    if (len(ds.partition_key) != 1 or
+            not isinstance(ds[ds.partition_key[0]].dtype, TLocus)):
```

Comment:
Is this code repeated in other places? We should probably make it a function to call similar to `require_biallelic`.

Reply:
currently just here. But I can pull it out just in case we want to use it elsewhere.